### ISSUE TITLE
[question] reproduce results?

### ISSUE THREAD

**Issue by jarlva** [alex-petrenko/sample-factory#65]
Hi, 
I find that reproducing results helps a lot while tweaking the various parameters by removing the randomness factor. Is it possible to reproduce the results?
I used the argument `--seed=5` in train_gym.py. Yet it does not lock the results. I get different results on tensorboard and also different completion times. So it's harder to pin down good parameters/code. The gym I use is reproducible in another agent).

Note the completion time difference for the two runs below.

![image](https://user-images.githubusercontent.com/26058035/95654781-00267c00-0b0b-11eb-86e5-a13a77ba5f74.png)

**Comment by alex-petrenko** [comment]
Hi, sorry for the delay, I missed the notification somehow.

The seed parameter fixes the initial set of weights of your policy. After that, there is no guarantee that the run will be exactly reproducible. There are lots of other computations that are extremely hard to make fully reproducible, so most frameworks won't even try (because it's 99.9% futile unless you make everything fully synchronous and sequential).

The main source of randomness in Sample Factory is action sampling from the stochastic policy. It is not bootstrapped from the seed AFAIK.
But in general, parallel computations are inherently not reproducible in most circumstances. It is almost impossible to control in which order the data will arrive at the learner from multiple actor workers. And the order of data will change the gradients.

I don't know of any other async RL framework that can provides precise reproducibility, and most synchronous frameworks won't guarantee it as well. I'd be curious to know what framework you're using such that the runs are exactly reproducible.

The last point: precise repeatability of experiments can actually harm your research. You can overfit hyperparameters to your set of initial weights and to other conditions, and the slightest change to the agent or environment will harm the performance.
A better thing to do in RL is to run multiple seeds, estimate mean and standard deviation and then act based on that. E.g. if the change you made is still within one standard deviation of the natural randomness, you probably haven't improved your algorithm.

**Comment by alex-petrenko** [comment]
Some more info on that in PyTorch docs: https://pytorch.org/docs/stable/notes/randomness.html
E.g. it's possible to make some of the computations deterministic in CuDNN, but it harms performance (and it still won't help with an async parallel system).

**Comment by jarlva** [comment]
Yes, after the question I found that link..
I'm aware that it's important to explore as much as possible. Yet, to compare apples-to-apples it saved me tons of times.
To your question, stable baseline(1) is reproducible. https://stable-baselines.readthedocs.io/en/master/guide/algos.html#reproducibility 

p.s. would it be possible in appo.py to remove tensorbaordX dependency by:
replace line: `from tensorboardX import SummaryWriter`
with: `from torch.utils.tensorboard import SummaryWriter`

It's also able to display white space in labels instead of underline _

**Comment by alex-petrenko** [comment]
Thank you for the link @jarlva. It mostly confirms what I said earlier, it is very hard to achieve reproducibility with parallel computations. They require you to use only CPU training and only use one core for training to guarantee reproducibility, which will make the training very slow. GPU reproducibility is not supported.
If you absolutely need reproducibility for your research, perhaps SF is a wrong framework to use.

I remember discussing the tensorboardX issue, I think yes, it should be possible to use torch.utils.tensorboard now. Not a very high priority, because the current solution works too. Is dependency on tensorboardX causing issues for you?

If replacing tensorboardX with torch.tensorboard can preserve all of the existing functionality, you're welcome to submit a PR!
